src/geo_traits_impl.rs

```diff
@@ -0,0 +1,717 @@
+use std::hint::unreachable_unchecked;
+
+use crate::{
+    Multipoint, MultipointM, MultipointZ, Point, PointM, PointZ, PolygonRing, Polyline, PolylineM,
+    PolylineZ, NO_DATA,
+};
+use geo_traits::{
+    CoordTrait, LineStringTrait, MultiLineStringTrait, MultiPointTrait, MultiPolygonTrait,
+    PointTrait, PolygonTrait,
+};
+
+// Shapefile points can't be null, so we implement both traits on them
+impl CoordTrait for Point {
+    type T = f64;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        geo_traits::Dimensions::Xy
+    }
+
+    fn nth_or_panic(&self, n: usize) -> Self::T {
+        match n {
+            0 => self.x(),
+            1 => self.y(),
+            _ => panic!("invalid dimension index"),
+        }
+    }
+
+    unsafe fn nth_unchecked(&self, n: usize) -> Self::T {
+        match n {
+            0 => self.x(),
+            1 => self.y(),
+            _ => unreachable_unchecked(),
+        }
+    }
+
+    fn x(&self) -> Self::T {
+        self.x
+    }
+
+    fn y(&self) -> Self::T {
+        self.y
+    }
+}
+
+impl CoordTrait for &Point {
+    type T = f64;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        geo_traits::Dimensions::Xy
+    }
+
+    fn nth_or_panic(&self, n: usize) -> Self::T {
+        match n {
+            0 => self.x(),
+            1 => self.y(),
+            _ => panic!("invalid dimension index"),
+        }
+    }
+
+    unsafe fn nth_unchecked(&self, n: usize) -> Self::T {
+        match n {
+            0 => self.x(),
+            1 => self.y(),
+            _ => unreachable_unchecked(),
+        }
+    }
+
+    fn x(&self) -> Self::T {
+        self.x
+    }
+
+    fn y(&self) -> Self::T {
+        self.y
+    }
+}
+
+impl PointTrait for Point {
+    type T = f64;
+    type CoordType<'a> = &'a Point where Self: 'a;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        geo_traits::Dimensions::Xy
+    }
+
+    fn coord(&self) -> Option<Self::CoordType<'_>> {
+        Some(self)
+    }
+}
+
+impl PointTrait for &Point {
+    type T = f64;
+    type CoordType<'a> = &'a Point where Self: 'a;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        geo_traits::Dimensions::Xy
+    }
+
+    fn coord(&self) -> Option<Self::CoordType<'_>> {
+        Some(self)
+    }
+}
+
+// Shapefile points can't be null, so we implement both traits on them
+impl CoordTrait for PointM {
+    type T = f64;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        // TODO: is it valid to have NO_DATA here? I would've expected that a `PointM` would
+        // _always_ have an M coordinate.
+        // if self.m <= NO_DATA {
+        //     geo_traits::Dimensions::Xy
+        // } else {
+        //     geo_traits::Dimensions::Xym
+        // }
+        geo_traits::Dimensions::Xym
```

Comment:
This is not consistent with the import for &PointM

Reply:
I updated this to check if `self.m <= NO_DATA`, but I'm not sure if that's the correct behavior. It seems like a `PointM` should _always_ have an `m` value, because if it didn't have one, it could just be a `Point`, right?